File: app/src/main/java/com/vonage/tutorial/phonetoappabdul/outgoing/CallConnectionService.java
package com.vonage.tutorial.phonetoappabdul.outgoing;

import android.annotation.TargetApi;
import android.net.Uri;
import android.os.Build;
import android.telecom.Connection;
import android.telecom.ConnectionRequest;
import android.telecom.ConnectionService;
import android.telecom.PhoneAccountHandle;
import android.telecom.TelecomManager;
import android.util.Log;

import androidx.annotation.RequiresApi;

@TargetApi(Build.VERSION_CODES.M)
public class CallConnectionService extends ConnectionService {

    private static final String TAG = CallConnectionService.class.getSimpleName();

    public CallConnectionService() {
        super();
    }

    @Override
    public void onCreateIncomingConnectionFailed(
            PhoneAccountHandle connectionManagerPhoneAccount,
            ConnectionRequest request
    ) {
        super.onCreateIncomingConnectionFailed(connectionManagerPhoneAccount, request);
        Log.i(TAG, "onCreateIncomingConnectionFailed");
    }

    @TargetApi(Build.VERSION_CODES.N_MR1)
    @Override
    public Connection onCreateIncomingConnection(
            PhoneAccountHandle connectionManagerPhoneAccount,
            ConnectionRequest request
    ) {
        super.onCreateIncomingConnection(connectionManagerPhoneAccount, request);

        Log.i(TAG, "onCreateIncomingConnection");

        CallConnection conn = new CallConnection(this);
        conn.processCall("inbound");

        conn.setConnectionProperties(Connection.PROPERTY_SELF_MANAGED);
        conn.setAddress(Uri.parse("tel:" + "+00000000000"), TelecomManager.PRESENTATION_ALLOWED);
        conn.setInitializing();
        return conn;
    }

    @Override
    public void onCreateOutgoingConnectionFailed(
            PhoneAccountHandle connectionManagerPhoneAccount,
            ConnectionRequest request
    ) {
        super.onCreateOutgoingConnectionFailed(connectionManagerPhoneAccount, request);
        Log.i(TAG, "onCreateOutgoingConnectionFailed");
    }

    @TargetApi(Build.VERSION_CODES.O)
    @Override
    public Connection onCreateOutgoingConnection(
            PhoneAccountHandle connectionManagerPhoneAccount,
            ConnectionRequest request
    ) {
        super.onCreateOutgoingConnection(connectionManagerPhoneAccount, request);

        Log.i(TAG, "onCreateOutgoingConnection");

        CallConnection conn = new CallConnection(this);
        conn.processCall("outbound");

        conn.setConnectionProperties(Connection.PROPERTY_SELF_MANAGED);
        conn.setAddress(Uri.parse("sip:" + "andriy"), TelecomManager.PRESENTATION_ALLOWED);
        conn.setInitializing();
        return conn;
    }
}

File: app/src/main/java/com/vonage/tutorial/phonetoappabdul/CallConnectionService.kt
//package com.vonage.tutorial.phonetoappabdul
//
//import android.net.Uri
//import android.os.Build
//import android.telecom.*
//import androidx.annotation.RequiresApi
//import com.google.firebase.messaging.RemoteMessage
//import org.json.JSONObject
//
//class CallConnectionService : ConnectionService() {
//
//    @RequiresApi(Build.VERSION_CODES.O)
//    override fun onCreateIncomingConnection(
//        connectionManagerPhoneAccount: PhoneAccountHandle?,
//        request: ConnectionRequest?
//    ): Connection {
//        /*
//        CSDemo: This gets the push info from the ClientManager and pulls out the from number as specified in your NCCO.
//        A CallConnection Object is also created. This is how the system tells you the user has initiated an action with the System UI.
//         */
//        val pushInfoBytes = request?.extras?.getByteArray("pushinfo")
//        val pushInfo = ParcelableUtil.unmarshall(pushInfoBytes!!, RemoteMessage.CREATOR)
//        val connection = CallConnection(this, pushInfo)
//        connection.processCall()
//        val from =
//            (((JSONObject(pushInfo.data["nexmo"]).get("body") as JSONObject).get("channel") as JSONObject).get(
//                "from"
//            ) as JSONObject).get("user") as String
//        connection.setCallerDisplayName(from, TelecomManager.PRESENTATION_ALLOWED)
//        connection.setAddress(Uri.parse("tel: Olx calling..."), TelecomManager.PRESENTATION_ALLOWED)
//        connection.connectionProperties = Connection.PROPERTY_SELF_MANAGED
//        connection.setInitializing()
//        return connection
//    }
//
//    override fun onCreateIncomingConnectionFailed(
//        connectionManagerPhoneAccount: PhoneAccountHandle?,
//        request: ConnectionRequest?
//    ) {
//        super.onCreateIncomingConnectionFailed(connectionManagerPhoneAccount, request)
//    }
//}